### ISSUE TITLE
Line cursor width does not change

### ISSUE THREAD

**Issue by yamam** [mintty/mintty#1157]
Changing the cursor width in the windows accessibility options does not change the width of the mintty line cursor.
The caret_width that can be obtained by the following process in wintext.c changes.
```c
SystemParametersInfo(SPI_GETCARETWIDTH, 0, &caret_width, 0);
```
However, since line_width is set to 1, caret_width is set to 1 in the following process.
```c
        if (caret_width > line_width)
          caret_width = line_width;
```
line_width is calculated by: attr.attr was 0x4000030180 lattr was 0 and cell_height was 17.
``` c
  int line_width = (3
                    + (attr.attr & ATTR_BOLD ? 1 : 0)
                    + (attr >= LATTR_WIDE ? 2 : 0)
                    + (lattr >= LATTR_TOP ? 2 : 0)
                   ) * cell_height / 40;
```
I am using mintty eac7c0bfa7f4a116221564306d747f09c5656750 in wsltty.
I checked wsltty/config with empty.

**Comment by mintty** [comment]
An easy solution would be to use the InvertRect function, see attached screenshot. Drawback is that we have no control over the colours used for the cursor.
![inverted-line-cursor](https://user-images.githubusercontent.com/12740416/170237108-a9bb8cdc-d1a9-4b03-81b3-f1c47554789f.png)

**Comment by yamam** [comment]
I read #1011 and understand why the line cursor width is limited.
Here is a screenshot with widths 1 and 2 (I changed the mintty source code to achieve this). I prefer to use width 2 because it is difficult to see where the cursor is on the 1 width.
![1](https://user-images.githubusercontent.com/12369962/170381614-e393334a-b1b9-480b-8bb1-3d6e9bfe3b0e.png)
![2](https://user-images.githubusercontent.com/12369962/170381630-6352cd00-f0b4-49d9-a2a5-7b07974241fc.png)

In Windows notepad, the text under the cursor seems to be inverted. mintty and gvim do not seem to invert the cursor. But with a width of 2, I don't see a problem with no inverted display.
![notepad3](https://user-images.githubusercontent.com/12369962/170381684-805c9c62-7452-47f7-adb1-00b164339adf.png)

I would like to be able to set the width of the line cursor in the mintty config. Also, I would like to see the restriction by line_width removed. Setting the line cursor width to 1 would be the same as the current behavior and would solve the problem #1011.

**Comment by ak2** [comment]
This is also an issue with high-dpi displays, where it's quite easy to lose
track of a single-pixel–line cursor. I think it would be good to go to two
pixels for 200% scaling (and more for higher scaling).

**Comment by mintty** [comment]
Please test the uploaded change.

**Comment by yamam** [comment]
Thanks for the fix.
I could confirm that I can change the thickness of the cursor in mintty by changing the text cursor thickness setting in windows.
However, in the April 24 version, the color of the text under the cursor turned white and the text was visible, but in the latest version, the color of the text under the cursor did not change, making the text difficult to read.
![image](https://user-images.githubusercontent.com/12369962/194672761-1fc9be5d-db7e-45a3-a1e5-0ffa8f731989.png)
![image](https://user-images.githubusercontent.com/12369962/194672681-61d227d0-d590-420f-8a9b-c35a6af0907d.png)

**Comment by mintty** [comment]
Presviously, the line cursor would overlay the character, thus making it invisible. I changed that towards normal cursor behavioiur, like the full block cursor. You should change cursor colour to be sufficiently distinct from text colour.

**Comment by yamam** [comment]
Why did you choose to operate that way?
I prefer a blue cursor, but it is very difficult to see if the text is green or blue.
If the text under the cursor was white, as it was before, it would be easier to read regardless of the color of the text.

**Comment by mintty** [comment]
I've applied some cursor contrast enforcement heuristics to all cursor styles. Please retest.
Also please confirm that high-DPI cursor handling is now hopefully better.

**Comment by yamam** [comment]
Thank you for your prompt response.
However, there appears to be no change.
![image](https://user-images.githubusercontent.com/12369962/194707333-785c7ac4-8334-4a13-ba84-e0c8ff168152.png)
I've checked the default settings as well.
The text under the cursor is hard to see.
![image](https://user-images.githubusercontent.com/12369962/194707312-f59e8cdb-d018-4374-879f-0f0594a49484.png)

**Comment by mintty** [comment]
This is basically a consequence of your choice of cursor colour very close to text colour, which is not a good idea in the first place. But if you send me your effective colour values, I'll look into it once more.
What about high-DPI cursor width behaviour?

**Comment by mintty** [comment]
OK, looking again, that's not default foreground colour but colour attribute setting where the cursor is obscured.
Anyway, I'd appreciate the values as a test case.

**Comment by yamam** [comment]
> This is basically a consequence of your choice of cursor colour very close to text colour, which is not a good idea in the first place. But if you send me your effective colour values, I'll look into it once more.

No matter what color the cursor is, text editors such as vim will display text of various colors, so text similar to the cursor color will be difficult to see.

I usually use the following settings.
```
ForegroundColour=64,64,64
BackgroundColour=255,244,232
CursorType=block
CursorBlinks=no
CursorColour=0,128,240
```
![image](https://user-images.githubusercontent.com/12369962/194707965-c6116e88-68a9-4b29-9f74-71c8fa46e479.png)

And an empty config file will display the following colors.
![image](https://user-images.githubusercontent.com/12369962/194708077-5a6f1a02-cfb2-4b4d-a7d3-964a358da746.png)

**Comment by yamam** [comment]
> What about high-DPI cursor width behaviour?

Sorry, I don't have a high-dpi environment, so I can't check.

**Comment by mintty** [comment]
Now this is confusing me; this issue was about the line style cursor, now you've configured CursorType=block.
Also, for me, an empty config file, changing cursor type to block, makes it look like this:
![grafik](https://user-images.githubusercontent.com/12740416/194709028-0cf6ef92-4efa-4284-b469-e9de302a03d0.png)
About the green/green contrast, can you determine which shade that editor uses?
About DPI, that was someone else's comment, sorry.

**Comment by yamam** [comment]
> Now this is confusing me; this issue was about the line style cursor, now you've configured CursorType=block.

I use both line and block cursors, line cursor in vim insert mode and block cursor in normal mode.

> Also, for me, an empty config file, changing cursor type to block, makes it look like this:

The text in your screenshot is legible because the color of the cursor is different from the color of the text. I don't know why the cursor and text color is the same in my environment.

> About the green/green contrast, can you determine which shade that editor uses?

CursorColour=0,128,240
BackgroundColour=255,244,232
The green letters are the following colors.
#00884c (0, 136, 76)

**Comment by mintty** [comment]
I've reverted the contrast tweaking heuristics as it was broken on block cursor style.
Contrast tweaking for line/underscore cursor styles is not implemented and not considered important as those cursors should normally be thin enough to recognize the character.
The fix remains to underlay, not overlay the character with the line cursor, and to make its width variable.

**Comment by yamam** [comment]
Thanks for the correction.
Is it possible to change the width of the line cursor by mintty (wsltty) config file or escape sequence?
I would like to change the width of the line cursor in mintty without changing the windows text cursor settings if possible.

**Comment by mintty** [comment]
https://github.com/mintty/mintty/wiki/CtrlSeqs#cursor-style
I thought as the dynamic configuration sequence already exists, static configuration may be dispensible.

**Comment by mintty** [comment]
Released 3.6.2.
